### PR TITLE
James 1784 Add authentication when downloading an attachment

### PR DESCRIPTION
**Pull request by aduprat**
Only the last commit has to be reviewed, others are located:
- JAMES-1777: #301 
- JAMES-1783: #302

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/AttachmentAccessToken.java [hunk 1]
```diff
@@ -98,7 +113,9 @@ public String serialize() {
     
     @Override
     public String getContent() {
-        return DateTimeFormatter.ISO_OFFSET_DATE_TIME.format(expirationDate);
+        return blobId
```
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/AttachmentAccessToken.java]: This modification is temporary, as #302 will include some changes on that.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java [hunk 2]
```diff
@@ -76,50 +86,99 @@ public void appendMessageWithAttachmentToMailbox(String user, String mailbox, St
...
-        response = with().options("/download/myBlob");
+        response = with()
+                .options("/download/" + KNOWN_BLOB_ID);
     }
 
     @When("^\"([^\"]*)\" downloads \"([^\"]*)\"$")
     public void downloads(String username, String attachmentId) throws Throwable {
         String blobId = blobIdByAttachmentId.get(attachmentId);
         AccessToken accessToken = userStepdefs.tokenByUser.get(username);
         if (accessToken != null) {
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: Can't it be done in more readable wa with optional ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java [hunk 3]
```diff
@@ -76,50 +86,99 @@ public void appendMessageWithAttachmentToMailbox(String user, String mailbox, St
...
+            response = with()
+                    .get("/download/" + blobId);
         }
+    }
+
+    private String attachmentAccessToken(String blobId, AccessToken accessToken) {
+        return with()
+            .header("Authorization", accessToken.serialize())
+            .post("/download/" + blobId)
+            .then()
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: indent here

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/AttachmentAccessToken.java [hunk 4]
```diff
@@ -33,7 +37,18 @@
...
         return new Builder();
     }
-    
+
+    public static AttachmentAccessToken from(String serializedAttachmentAccessToken) {
+        Preconditions.checkArgument(!Strings.isNullOrEmpty(serializedAttachmentAccessToken), "'AttachmentAccessToken' is mandatory");
+        List<String> split = Splitter.on(SEPARATOR).splitToList(serializedAttachmentAccessToken);
+        Preconditions.checkArgument(split.size() == 3, "Wrong 'AttachmentAccessToken'");
+        return builder()
+                .blobId(Iterables.get(split, 0, null))
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/AttachmentAccessToken.java]: What is null ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 2 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: Same comment as Raphael in the previous review.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 3 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: OK

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/AttachmentAccessToken.java (hunk 4 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/AttachmentAccessToken.java]: OK

**Comment by aduprat** [comment]
#306

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature [hunk 5]
```diff
@@ -17,35 +17,58 @@ Feature: Download endpoint
     When "usera@domain.tld" checks for the availability of the attachment endpoint
     Then the user should be authorized
 
+  Scenario: An unauthenticated user should not have access to the download endpoint
+    When "usera@domain.tld" downloads "a1"
+    Then the user should not be authorized
+
+  Scenario: An authenticated user should not have access to the download endpoint without the authentication token
+    Given "usera@domain.tld" is connected
+    When "usera@domain.tld" downloads "a1" without the authentication token
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature]: s/the/any/

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature [hunk 6]
```diff
@@ -17,35 +17,58 @@ Feature: Download endpoint
...
+    Then the user should not be authorized
+
+  Scenario: An authenticated user should not have access to the download endpoint with an empty authentication token
+    Given "usera@domain.tld" is connected
+    When "usera@domain.tld" downloads "a1" with empty authentication token
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature]: s/empty/an empty/

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java [hunk 7]
```diff
@@ -40,7 +40,8 @@
     @Inject
     private JMAPServer(JMAPConfiguration jmapConfiguration,
                        AuthenticationServlet authenticationServlet, JMAPServlet jmapServlet, DownloadServlet downloadServlet,
-                       AuthenticationFilter authenticationFilter, FirstUserConnectionFilter firstUserConnectionFilter) {
+                       AuthenticationFilter authenticationFilter, FirstUserConnectionFilter firstUserConnectionFilter,
+                       AttachmentAuthenticationFilter queryParameterAuthenticationFilter) {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java]: weird name

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature (hunk 5 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature]: OK, in #306

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature (hunk 6 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadEndpoint.feature]: OK, in #306

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java (hunk 7 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java]: OK, in #306

**Pull request closed without merging**